pkg/controller/deployment/sync.go

```diff
@@ -549,3 +550,15 @@ func (dc *DeploymentController) isScalingEvent(ctx context.Context, d *apps.Depl
 	}
 	return false, nil
 }
+
+// generateReplicaSetName generates a ReplicaSet name by adding a pod template spec hash
+// to a Deployment name, optionally truncating it to ensure the ReplicaSet name is within the limit.
+func generateReplicaSetName(deploymentName, podTemplateSpecHash string) string {
+	maxDeploymentNameLength := validation.DNS1123SubdomainMaxLength - 1 - len(podTemplateSpecHash)
+
+	if len(deploymentName) <= maxDeploymentNameLength {
+		return deploymentName + "-" + podTemplateSpecHash
+	}
+
+	return deploymentName[:maxDeploymentNameLength] + "-" + podTemplateSpecHash
```

Comment:
nit: it's slightly more readable to put the special case (deployment name too long) in the `if` and return the common case. Suggest swapping the returns.